dfa: Two phase DFA

fridis opened the pull request:
Run DFA in two phases: First, to collect calls in `CallGroups` with common call destination, call site and call target to find all environments required by this call group.

Second, to run the analysis tracing effect environments as well, but limiting this to the environments that are actually required as found in the first phase for the group each call belongs to.

For this to work, the following changes were needed:

- `effect.instate0` is now called by `instate_helper` such that distinguishing the `instate_helper` instances is no longer needed to distinguish the effect environments
- `Java_String.utf8` no longer uses `fuzion.jvm.env` since the analysis is no longer accurate to trace the presence of the `fuzion.jvm` effect. This may lead to errors at runtime if a `Java_String` is used outside of the jvm effect
- added `FUIR.clazzDeclarationPos()` for better error output for Environments, implemented `FUIR.clazzSrcFile` using `FUIR.clazzDeclarationPos()`.
- added new class `CallGroup` to goup calls with same environment
- In a first phase, all the `CallGroup` and the effects that are used or possibly available in each `CallGroup` are determined. Then, DFA is restarted from scratch adding effect environments to calls, but only for those effects that are both used and possibly available by the corresponding `CallGroup`. 
- A new env var / property `dev.flang.fuir.analysis.dfa.TRACE_ALL_EFFECT_ENVS` can be used to enforce full effect environments for all calls independent on what was found in the `CallGroup`
- A test `tests/tricky_dfa_cases` was added that contains cases that required full effect environments.

fridis committed:
dfa: Two phase DFA

Run DFA in two phases: First, to collect calls in `CallGroups` with common call
destination, call site and call target to find all environments required by this
call group.

Second, to run the analysis tracing effect environments as well, but limiting
this to the environments that are actually required as found in the first phase
for the group each call belongs to.

fridis committed:
tests: skipped test and updated err output

fridis committed:
base: move call to `ìnstate0` to `instate_helper`

fridis committed:
dfa: Join instances into a single instance

fridis committed:
tests: update expected error output due to changed stack trace

fridis committed:
dfa: propagate needed effects to targets of calls

This helps DFA on tricky cases that I am adding in a new test

fridis committed:
base: workaround for DFA analysis that currently fails for `Java_String.utf8`

The problem is that the `Java_String` instances apparently flow to code that is
executed outside of the `fuzion.jvm` effect, need to check why this his the
case.  Small example to reproduce:

    q(s String) => _ := s+(codepoint 10)
    q "a"
    fuzion.jvm.use ()->
      q Java.java.lang.__jString.new

fridis committed:
tests: Add missing module `java.base` to `tricky_dfa_cases`

fridis committed:
dfa: disable propagation of needed effects to callgroup target

fridis committed:
tests: disable tricky_dfa_cases, currently does not work

fridis committed:
Merge branch 'main' into two_phase_dfa

fridis committed:
base: cosmetics: remove extra blank line

fridis committed:
dfa: cleanup

fridis committed:
tests: update expected err output

fridis committed:
tests: update expected output for effects

just the order changed

fridis committed:
Merge branch 'main' into two_phase_dfa

fridis committed:
fuir: implemented `FUIR.clazzSrcFile` using `FUI.clazzDeclarationPos`

fridis committed:
util: Fix NullPointer in precondition failuer in `SourcePosition`

This crashed if created from a `SourceFile` before `_bytes` was set.

fridis committed:
util: Fix javac compilation due to repeated init of final field

fridis committed:
fuir: re-introduce `LibraryFuir.clazzSrcFile`, crashes on MacOS without this

fridis committed:
dfa: cleanup: remove duplicate fields, add methods, add comments, etc.

fridis committed:
dfa: cleanup: join getEffectCheck/getEffectCheck2/getEffectForce into 1 method

fridis committed:
dfa: fixes in join getEffectCheck/getEffectCheck2/getEffectForce into 1 method

No errror must be reported in case of `effect.type.is_instated0`.

fridis committed:
dfa: fixes in join getEffectCheck/getEffectCheck2/getEffectForce into 1 method

Do not require effect to be instated for calls to `effect.type.unsafe_from_env`.

fridis committed:
tests: update expected effects

fridis committed:
dfa: Cleanup in CallGroup, removed debug code, add comments, etc.

fridis committed:
Merge branch 'main' into two_phase_dfa

fridis committed:
dfa: Cleanup in CallGroup, renamed methods, addec comments, removed dead code

fridis committed:
dfa: Cleanup: remove not needed code

fridis committed:
dfa: Cleanup: comments, move to better suited section, remove redundant field

fridis committed:
Merge branch 'main' into two_phase_dfa

fridis committed:
dfa: cleanup: comments, code structure, added NYI-comments

fridis committed:
dfa: cleanup: removed not needed code.

fridis committed:
util: add comment to explain why dummy `SourceFile` instance is created

fridis committed:
dfa: Add `_site` back to `Call` since `CallGroup._site` too unspecific

Using the site from the CallGroup results in error messages showing sites that
are actually not problematic, so this patch re-introduces the site to `Call` for more detail.

In particular, this allows `tests/effect_installed_negative` to be re-enabled
after a slight change.

fridis committed:
tests: update expected err, call site where effect is missing is correct again

fridis committed:
dfa: Add property/envvar to re-enable full accurary for effect environments

This enables running tests/tricky_dfa_tests.

fridis committed:
tests: Update expected err output due to `To fix this...` message that was added

fridis committed:
tests: Update expected err output due to typo

maxteufel reviewed (changes requested):
Putting a blocker on this, as discussed per internal mailing list. This pull request currently increases the duration of the DFA for the webserver.

fridis committed:
dfa: fix performance degradation from commit b874f17

The problem was that `effect.type.unsafe_from_env` marked the effect as `used`
resulting in additional environments created during DFA.

This patch renames `Call.getEffectCheck` as `Call.useAndGetEffect` and adds a
new method `Call.getEffect`that does not mark the effect as used.

maxteufel reviewed (approved):
Performance is now fixed in the webserver again!

fridis committed:
dfa: re-enable verbose output during DFA iterations

verbose level 2 to shoulw each iteration

verbose level 3 to also show number of calls, values and envs

verbose level 4 to show reason why fix point is not yet reached

fridis committed:
Merge branch 'main' into two_phase_dfa

fridis committed:
Merge branch 'main' into two_phase_dfa
